File: tools/atlas-parser/atlas-parser.ts
import {
  type Anim,
  type Atlas,
  animCels,
  animMillis,
  celMillis,
  type TagFormat
} from '../../src/graphics/atlas.ts'
import {type Box, boxEq, type XY} from '../../src/types/geo.ts'
import type {
  Aseprite,
  AsepriteDirection,
  AsepriteFrame,
  AsepriteFrameMap,
  AsepriteFrameTag,
  AsepriteSlice,
  AsepriteTagSpan
} from './aseprite.ts'

export function parseAtlas(ase: Aseprite): Atlas {
  const anim: {[tag: string]: Anim} = {}
  const cels: number[] = []
  for (const span of ase.meta.frameTags) {
    const tag = parseTag(span.name)
    if (anim[tag]) throw Error(`atlas tag "${tag}" duplicate`)
    const id = Object.keys(anim).length
    anim[tag] = parseAnim(id, span, ase.frames, ase.meta.slices)
    for (const cel of [...parseAnimFrames(span, ase.frames)].map(parseCel))
      cels.push(cel.x, cel.y, anim[tag].w, anim[tag].h)
  }
  for (const slice of ase.meta.slices)
    if (!anim[parseTag(slice.name)])
      throw Error(`atlas hitbox "${slice.name}" has no animation`)
  return {anim, cels, tags: Object.keys(anim)}
}

/** @internal */
export function parseAnim(
  id: number,
  span: AsepriteTagSpan,
  map: AsepriteFrameMap,
  slices: readonly AsepriteSlice[]
): Anim {
  const frame = parseAnimFrames(span, map).next().value
  if (!frame) throw Error(`no atlas frame "${span.name}"`)
  const {hitbox, hurtbox} = parseHitboxes(span, slices)
  return {
    cels: span.to - span.from + 1,
    h: frame.sourceSize.h,
    hitbox,
    hurtbox,
    id,
    w: frame.sourceSize.w
  }
}

/** @internal */
export function* parseAnimFrames(
  span: AsepriteTagSpan,
  map: AsepriteFrameMap
): IterableIterator<
  AsepriteFrame,
  // biome-ignore lint/suspicious/noConfusingVoidType:;
  AsepriteFrame | void
> {
  let cels = 0
  let animDuration = 0
  const len = span.to - span.from + 1
  const peak = len - 1
  const cycle = Math.max(1, 2 * peak)
  const indexByDir: {[dir in AsepriteDirection]: (i: number) => number} = {
    forward: i => span.from + (i % len),
    pingpong: i => span.from + peak - Math.abs((i % cycle) - peak),
    pingpong_reverse: i => span.to - (peak - Math.abs((i % cycle) - peak)),
    reverse: i => span.to - (i % len)
  }
  const frameIndex = indexByDir[span.direction as AsepriteDirection]
  for (let i = 0; cels < animCels && animDuration < animMillis; i++) {
    const frameTag = `${span.name}--${frameIndex(i)}` as AsepriteFrameTag
    const frame = map[frameTag]
    if (!frame) throw Error(`no atlas frame "${frameTag}"`)
    for (
      let celDuration = 0;
      celDuration < frame.duration &&
      cels < animCels &&
      animDuration < animMillis;
      celDuration += celMillis, animDuration += celMillis, cels++
    )
      yield frame
  }
}

/** @internal */
export function parseCel(frame: AsepriteFrame): XY {
  return {
    x: frame.frame.x + (frame.frame.w - frame.sourceSize.w) / 2,
    y: frame.frame.y + (frame.frame.h - frame.sourceSize.h) / 2
  }
}

/** @internal */
export function parseHitboxes(
  span: AsepriteTagSpan,
  slices: readonly AsepriteSlice[]
): {hitbox: Box | undefined; hurtbox: Box | undefined} {
  let hitbox
  let hurtbox
  // https://github.com/aseprite/aseprite/issues/3524
  for (const slice of slices) {
    if (slice.name !== span.name) continue

    if (!slice.keys[0]) continue

    for (const k of slice.keys)
      if (!boxEq(k.bounds, slice.keys[0].bounds))
        throw Error(
          `atlas tag "${span.name}" hitbox bounds varies across frames`
        )

    const red = slice.color === '#ff0000ff'
    const green = slice.color === '#00ff00ff'
    const blue = slice.color === '#0000ffff' // default Aseprite slice color.
    if (!red && !green && !blue)
      throw Error(
        `atlas tag "${span.name}" hitbox color ${slice.color} unsupported`
      )

    if (hitbox && (red || blue))
      throw Error(`atlas tag "${span.name}" has multiple hitboxes`)

    if (hurtbox && (green || blue))
      throw Error(`atlas tag "${span.name}" has multiple hurtboxes`)

    if (red || blue) hitbox = slice.keys[0].bounds
    if (green || blue) hurtbox = slice.keys[0].bounds
  }
  return {hitbox, hurtbox}
}

function parseTag(tag: string): TagFormat {
  if (!tag.includes('--'))
    throw Error(`atlas tag "${tag}" not in <filestem>--<animation> format`)
  return tag as TagFormat
}

File: src/cmd/packsprites/atlas_parser.go
package main

import (
	"fmt"
	"strings"

	"github.com/oidoid/void/src/void/vatlas"
	"github.com/oidoid/void/src/void/vmath"
)

// converts raw Aseprite JSON into an Atlas and the ordered tag names. the
// returned tags slice is parallel to Atlas.Anims.
func parseAtlas(file *vatlas.AseFile) (*vatlas.Atlas, []string, error) {
	// void--Nil is always first at index 0 with a single transparent cel.
	anims := []vatlas.Anim{{Cels: 1}}
	celXY := []uint16{0, 0}
	tags := []string{"void--Nil"}
	seen := map[string]bool{"void--Nil": true}

	for _, span := range file.Meta.FrameTags {
		tag, err := parseTag(span.Name)
		if err != nil {
			return nil, nil, err
		}
		if seen[tag] {
			return nil, nil, fmt.Errorf("atlas tag %q duplicate", tag)
		}
		seen[tag] = true

		frames, err := parseAnimFrames(span, file.Frames)
		if err != nil {
			return nil, nil, err
		}
		anim, err := parseAnim(span.Name, frames, file.Meta.Slices)
		if err != nil {
			return nil, nil, err
		}
		anims = append(anims, anim)
		tags = append(tags, tag)
		for _, frame := range frames {
			xy := parseCel(frame)
			celXY = append(celXY, xy.X, xy.Y)
		}
	}

	for _, slice := range file.Meta.Slices {
		if !seen[slice.Name] {
			return nil, nil, fmt.Errorf("atlas hitbox %q has no animation", slice.Name)
		}
	}

	atlas := vatlas.NewAtlas(anims, celXY)
	return &atlas, tags, nil
}

func intAbs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func parseAnim(
	name string,
	frames []vatlas.AseFrame,
	slices []vatlas.AseSlice,
) (vatlas.Anim, error) {
	if len(frames) == 0 {
		return vatlas.Anim{}, fmt.Errorf("no atlas frame %q", name)
	}
	hitbox, hurtbox, err := parseHitboxes(name, slices)
	if err != nil {
		return vatlas.Anim{}, err
	}
	return vatlas.Anim{
		Cels:    uint8(len(frames)),
		W:       uint16(frames[0].SourceSize.W),
		H:       uint16(frames[0].SourceSize.H),
		Hitbox:  hitbox,
		Hurtbox: hurtbox,
	}, nil
}

// each frame is guaranteed to appear for at least `CelMillis`. cels are
// duplicated until cel duration is at least met; cels are unpacked until a full
// period is defined for the direction. no warns for overflowing past on second
// or uneven periods.
func parseAnimFrames(
	span vatlas.AseTagSpan,
	frameMap map[string]vatlas.AseFrame,
) ([]vatlas.AseFrame, error) {
	var frames []vatlas.AseFrame
	animDuration := 0
	from := int(span.From)
	to := int(span.To)
	frameLen := to - from + 1
	peak := frameLen - 1
	cycle := max(1, 2*peak)

	end := frameLen
	if span.Direction == "pingpong" || span.Direction == "pingpong_reverse" {
		end = cycle
	}

	indexByDir := func(i int) int {
		switch span.Direction {
		case "reverse":
			return to - (i % frameLen)
		case "pingpong":
			return from + peak - intAbs(i%cycle-peak)
		case "pingpong_reverse":
			return to - (peak - intAbs(i%cycle-peak))
		default: // forward
			return from + (i % frameLen)
		}
	}

	for i := 0; i < end &&
		len(frames) < vatlas.CelsPerAnim &&
		animDuration < vatlas.MaxAnimLoopMillis; i++ {
		// `--filename-format='{title}--{tag}--{frame}'`.
		frameTag := fmt.Sprintf("%s--%d", span.Name, indexByDir(i))
		frame, ok := frameMap[frameTag]
		if !ok {
			return nil, fmt.Errorf("no atlas frame %q", frameTag)
		}
		for celDur := 0; celDur < int(frame.Duration) &&
			len(frames) < vatlas.CelsPerAnim &&
			animDuration < vatlas.MaxAnimLoopMillis; celDur += vatlas.CelMillis {
			animDuration += vatlas.CelMillis
			frames = append(frames, frame)
			if from == to {
				return frames, nil // optimize for long single cel
			}
		}
	}
	return frames, nil
}

func parseCel(frame vatlas.AseFrame) vmath.XY[uint16] {
	x := frame.Frame.X + int32((frame.Frame.W-uint32(frame.SourceSize.W))/2)
	y := frame.Frame.Y + int32((frame.Frame.H-uint32(frame.SourceSize.H))/2)
	return vmath.NewXY(uint16(x), uint16(y))
}

func parseHitboxes(
	name string,
	slices []vatlas.AseSlice,
) (vmath.Box[uint16], vmath.Box[uint16], error) {
	var hitbox, hurtbox vmath.Box[uint16]
	for _, slice := range slices {
		if slice.Name != name || len(slice.Keys) == 0 {
			continue
		}
		first := slice.Keys[0].Bounds
		for _, k := range slice.Keys {
			if k.Bounds != first {
				return vmath.Box[uint16]{}, vmath.Box[uint16]{},
					fmt.Errorf(
						"atlas tag %q hitbox bounds varies across frames",
						name,
					)
			}
		}
		red := slice.Color == "#ff0000ff"
		green := slice.Color == "#00ff00ff"
		blue := slice.Color == "#0000ffff"
		if !red && !green && !blue {
			return vmath.Box[uint16]{}, vmath.Box[uint16]{},
				fmt.Errorf(
					"atlas tag %q hitbox color %s unsupported",
					name,
					slice.Color,
				)
		}
		var zero vmath.Box[uint16]
		if hitbox != zero && (red || blue) {
			return zero, zero, fmt.Errorf("atlas tag %q has multiple hitboxes", name)
		}
		if hurtbox != zero && (green || blue) {
			return zero, zero, fmt.Errorf("atlas tag %q has multiple hurtboxes", name)
		}
		box := vmath.XYWH(
			uint16(first.X),
			uint16(first.Y),
			uint16(first.W),
			uint16(first.H),
		)
		if red || blue {
			hitbox = box
		}
		if green || blue {
			hurtbox = box
		}
	}
	return hitbox, hurtbox, nil
}

func parseTag(name string) (string, error) {
	if !strings.Contains(name, "--") {
		return "", fmt.Errorf(
			"atlas tag %q not in <filestem>--<animation> format", name,
		)
	}
	return name, nil
}
